src/Events/QR/Routes.php

```diff
@@ -0,0 +1,257 @@
+	 *
+	 * @since TBD
+	 * @return void
+	 */
+	protected function add_hooks(): void {
+		add_action( 'tribe_events_pre_rewrite', [ $this, 'add_qr_rules' ] );
+		add_filter( 'query_vars', [ $this, 'filter_add_query_vars' ] );
+		add_filter( 'tribe_rewrite_parse_query_vars', [ $this, 'filter_parse_query_vars' ] );
+	}
+
+	/**
+	 * Removes the actions required by the controller.
+	 *
+	 * @since TBD
+	 * @return void
+	 */
+	protected function remove_hooks(): void {
+		remove_action( 'tribe_events_pre_rewrite', [ $this, 'add_qr_rules' ] );
+		remove_filter( 'query_vars', [ $this, 'filter_add_query_vars' ] );
+		remove_filter( 'tribe_rewrite_parse_query_vars', [ $this, 'filter_parse_query_vars' ] );
+	}
+
+	/**
+	 * Get the route prefix.
+	 *
+	 * @since TBD
+	 * @return string The route prefix.
+	 */
+	public function get_route_prefix(): string {
+		return $this->route_prefix;
+	}
+
+	/**
+	 * Add QR code rewrite rules.
+	 *
+	 * @since TBD
+	 * @param Tribe__Events__Rewrite $rewrite The TEC rewrite instance.
+	 * @return void
+	 */
+	public function add_qr_rules( Tribe__Events__Rewrite $rewrite ): void {
+		$rewrite->add(
+			[ 'events', $this->route_prefix, '([^/]+)' ],
+			[ 'tec_qr_hash' => '%1' ]
+		);
+	}
+
+	/**
+	 * Adds the required Query Vars for QR code routes.
+	 *
+	 * @since TBD
+	 * @param array $query_vars The array of query variables to add to.
+	 * @return array The modified query vars.
+	 */
+	public function filter_add_query_vars( $query_vars = [] ) {
+		$query_vars[] = 'tec_qr_hash';
+
+		return $query_vars;
+	}
+
+	/**
+	 * Parse query vars for QR code routes.
+	 *
+	 * @since TBD
+	 *
+	 * @param array $query_vars The current query vars.
+	 *
+	 * @return array The modified query vars.
+	 */
+	public function filter_parse_query_vars( array $query_vars ): array {
+		if ( isset( $query_vars['tec_qr_hash'] ) ) {
+			$query_vars['tec_qr_hash'] = sanitize_text_field( $query_vars['tec_qr_hash'] );
+		}
+
+		return $query_vars;
+	}
+
+	/**
+	 * Generate a unique hash for a QR code.
+	 *
+	 * @since TBD
+	 *
+	 * @param int    $post_id The post ID (event or series).
+	 * @param string $qr_type The QR Redirection behavior.
+	 *
+	 * @return string The generated hash.
+	 */
+	public function generate_hash( int $post_id, string $qr_type ): string {
+		// Create a simple string with the data and salt.
+		$data = $post_id . ':' . $qr_type . ':' . $this->salt;
+
+		// Convert to base64url (URL-safe base64).
+		return rtrim( strtr( base64_encode( $data ), '+/', '-_' ), '=' );
+	}
+
+	/**
+	 * Get the URL for a QR code.
+	 *
+	 * @since TBD
+	 *
+	 * @param int    $post_id The post ID (event or series).
+	 * @param string $qr_type The QR Redirection behavior.
+	 *
+	 * @return string The QR code URL.
+	 */
+	public function get_qr_url( int $post_id, string $qr_type ): string {
+		$hash = $this->generate_hash( $post_id, $qr_type );
+
+		return home_url( "events/{$this->route_prefix}/{$hash}/" );
```

Comment:
i dont like the hardcoded events ? Is it meant to direct the users to TEC's events archive ?